Milliehartnt123 update architectural overview (#31209)
backstage/backstage: docs/overview/architecture-overview.md

- Core - This includes the base functionality developed by core developers within the open-source project.
- App - The app represents a deployed instance of a Backstage application, customized and maintained by app developers, typically a productivity team within an organization. It integrates core functionalities with additional plugins.
- Plugins - These provide additional functionalities to enhance the usefulness of your Backstage app. Plugins can be company-specific or open-sourced and reusable.

## Overview

The following diagram shows a high level view of the overall architecture of Backstage. Running this architecture in a real environment typically involves
containerizing the components. Various commands are provided for accomplishing this.

There are 3 main components in this architecture:

> - The [frontend container](#frontend-building-blocks) includes the core Backstage [UI](#user-interface) which is an [extension](#extensions) that interacts directly with the user to present the information from the integrated core feature plugins, and other plugins added by a user.

Author: Changed "Frontend Container" to "Frontend" everywhere in the doc. Changed "Backend Container" to "Backend" everywhere in the doc.

## Plugin architecture

Architecturally, plugins can take three forms:

- [Standalone](#standalone-plugins)
- [Service backed](#service-backed-plugins)
- [Third-party backed](#third-party-backed-plugins)

### Standalone plugins

Standalone plugins run entirely in the browser. [The Tech Radar plugin](https://demo.backstage.io/tech-radar), for example, simply renders hard-coded information. It doesn't make any API requests to other services.

The architecture of the Tech Radar installed into a Backstage app is very simple. You just need to add Tech Radar as a frontend plugin into your app, as shown in the following diagram.

> **NOTE:**  
> The following diagram does not show the detailed contents of the frontend and backend containers in order to highlight the changes that pertain to the addition of the specified plugin.

![ui and tech radar plugin connected together](../assets/architecture-overview/simplified-standalone-plugin-architecture.jpeg)

Once the plugin has been added, then you can view the Tech Radar information in the Backstage UI.

![tech radar plugin ui](../assets/architecture-overview/tech-radar-plugin.png)

> ### Service backed plugins

Author: Made the change.

The Lighthouse plugin, for example, makes requests to the [lighthouse-audit-service](https://github.com/spotify/lighthouse-audit-service). The `lighthouse-audit-service` is a microservice which runs a copy of Google's [Lighthouse library](https://github.com/GoogleChrome/lighthouse/) and stores the results in a PostgreSQL database.

The Lighthouse plugin is added to the frontend. The lighthouse-audit-service container is already publicly available in Docker Hub and can be downloaded and run with

```bash
docker run spotify/lighthouse-audit-service:latest
```

> **NOTE:**  
> The following diagram does not show the detailed contents of the frontend and backend containers in order to highlight the changes that pertain to the addition of the specified plugin.

![lighthouse plugin backed to microservice and database](../assets/architecture-overview/simplified-service-based-plugin-architecture.jpeg)

The software catalog in Backstage is another example of a service backed plugin. It retrieves a list of services, or "entities", from the Backstage Backend service and renders them in a table for the user.

### Third-party backed plugins

> Third-party backed plugins are similar to service backed plugins. The main difference is that the service which backs the plugin is hosted outside of the ecosystem of the company hosting Backstage.

Reviewer: ```suggestion
Third-party backend plugins are similar to service backend plugins. The main difference is that the service which backs the plugin is hosted outside of the ecosystem of the company hosting Backstage.
```

> The CircleCI plugin is an example of a third-party backed plugin. CircleCI is a SaaS service which can be used without any knowledge of Backstage. It has an API which a Backstage plugin consumes to display content.

Author: Made the change

The Backstage CLI is in a category of its own and is depended on by virtually all other packages. It's not a library in itself though, and must always be a development dependency only.

### Deciding where you place your code

It can sometimes be difficult to decide where to place your plugin code. For example, should it go directly in the `-backend` plugin package or in the `-node` package? As a general guideline you should try to keep the exposure of your code as low as possible. If it doesn't need to be public API, it's best to avoid. If you don't need it to be used by other plugins, then keep it directly in the plugin packages.

Below is a chart to help you decide where to place your code.

![Package decision](../assets/architecture-overview/package-decision.drawio.svg)

## Cache

The Backstage backend and its built-in plugins are also able to leverage cache stores as a means of improving performance or reliability. Similar to how databases are supported, plugins receive logically separated cache connections, which are powered by [Keyv](https://github.com/lukechilds/keyv) under the hood.

> At this time of writing, Backstage can be configured to use one of five cache stores: memory, which is mainly used for local testing, memcache, redis, valkey or infinispan, which are cache stores better suited for production deployment. The right cache store for your Backstage instance will depend on your own run-time constraints and those required of the plugins you're running.

Reviewer: ```suggestion
At this time of writing, Backstage can be configured to use one of five cache stores: `memory`, `memcache`, `redis`, `valkey` or `infinispan`. `memory` is primarily used for local development, for production deployments, we recommend using one of the other cache stores. The right cache store for your Backstage instance will depend on your own run-time constraints and those required of the plugins you're running.
```

Author: Made the change. Changed the list of cache stores into a bulleted list.

Once the plugin has been added, then you can view the Tech Radar information in the Backstage UI.

![tech radar plugin ui](../assets/architecture-overview/tech-radar-plugin.png)

### Service backend plugins

Service backend plugins make API requests to a service which is within the purview of the organization running Backstage.

The Lighthouse plugin, for example, makes requests to the [lighthouse-audit-service](https://github.com/spotify/lighthouse-audit-service). The `lighthouse-audit-service` is a microservice which runs a copy of Google's [Lighthouse library](https://github.com/GoogleChrome/lighthouse/) and stores the results in a PostgreSQL database.

The Lighthouse plugin is added to the frontend. The lighthouse-audit-service container is already publicly available in Docker Hub and can be downloaded and run with

```bash
docker run spotify/lighthouse-audit-service:latest
```

> **NOTE:**  
> The following diagram does not show the detailed contents of the frontend and backend, in order to highlight the changes that pertain to the addition of the specified plugin.

> ![lighthouse plugin backed to microservice and database](../assets/architecture-overview/simplified-service-based-plugin-architecture.jpeg)

Reviewer: ```suggestion
![lighthouse plugin backend to microservice and database](../assets/architecture-overview/simplified-service-based-plugin-architecture.jpeg)
```

Author: @aramissennyeydd Made the change.

The software catalog in Backstage is another example of a service backend plugin. It retrieves a list of services, or "entities", from the Backstage Backend service and renders them in a table for the user.

### Third-party backend plugins

Third-party backend plugins are similar to service backend plugins. The main difference is that the service which backs the plugin is hosted outside of the ecosystem of the company hosting Backstage.

The CircleCI plugin is an example of a third-party backed plugin. CircleCI is a SaaS service which can be used without any knowledge of Backstage. It has an API which a Backstage plugin consumes to display content.

Requests going to CircleCI from the user's browser are passed through a proxy service that Backstage provides. Without this, the requests would be blocked by Cross Origin Resource Sharing policies which prevent a browser page served at [https://example.com](https://example.com) from serving resources hosted at https://circleci.com.

> **NOTE:**

> > The following diagram does not show the detailed contents of the frontend and backen, in order to highlight the changes that pertain to the addition of the specified plugin.

Reviewer: ```suggestion
> The following diagram does not show the detailed contents of the frontend and backend, in order to highlight the changes that pertain to the addition of the specified plugin.
```
